tools/codechecker_report_hash/codechecker_report_hash/hash.py

```diff
@@ -0,0 +1,325 @@
+    """ Generate report hash without bug path.
+
+    !!! NOT Compatible with the old hash generation method
+
+    High level overview of the hash content:
+     * 'file_name' from the main diag section.
+     * 'checker message'.
+     * 'line content' from the source file if can be read up. All the
+       whitespaces from the source content are removed.
+     * 'column numbers' from the main diag sections location.
+    """
+    try:
+        m_loc = diag.get('location')
+        source_line = m_loc.get('line')
+
+        from_col = m_loc.get('col')
+        until_col = m_loc.get('col')
+
+        # WARNING!!! Changing the error handling type for encoding errors
+        # can influence the hash content!
+        line_content = __get_line(source_file, source_line, errors='ignore')
+
+        # Remove whitespaces so the hash will be independet of the
+        # source code indentation.
+        line_content, new_col = remove_whitespace(line_content, from_col)
+
+        # Update the column number in sync with the
+        # removed whitespaces.
+        until_col = until_col - (from_col-new_col)
+        from_col = new_col
+
+        if line_content == '' and not os.path.isfile(source_file):
+            LOG.error("Failed to include soruce line in the report hash.")
+            LOG.error('%s does not exists!', source_file)
+
+        file_name = os.path.basename(source_file)
+        msg = diag.get('description')
+
+        hash_content = [file_name,
+                        msg,
+                        line_content,
+                        str(from_col),
+                        str(until_col)]
+
+        return __str_to_hash('|||'.join(hash_content))
+
+    except Exception as ex:
+        LOG.error("Hash generation failed")
+        LOG.error(ex)
+        return ''
+
+
+def get_report_path_hash(bug_path, files):
+    """ Returns path hash for the given bug path.
+
+    This can be used to filter deduplications of multiple reports.
+    """
+    report_path_hash = ''
+    events = filter(lambda i: i.get('kind') == 'event', bug_path)
+
+    for event in events:
+        file_name = os.path.basename(files[event['location']['file']])
+        line = str(event['location']['line']) if 'location' in event else 0
+        col = str(event['location']['col']) if 'location' in event else 0
+
+        report_path_hash += line + '|' + col + '|' + event['message'] + \
+            file_name
+
+    if not report_path_hash:
+        LOG.error('Failed to generate report path hash!')
+        LOG.error(bug_path)
+
+    LOG.debug(report_path_hash)
+    return __str_to_hash(report_path_hash)
+
+
+def use_report_hash(plist_file, version='v1'):
+    """ Override hash in the given file by using the given version hash. """
+    try:
+        plist = plistlib.readPlist(plist_file)
+
+        files = plist['files']
+
+        for diag in plist['diagnostics']:
+            file_path = files[diag['location']['file']]
+
+            if version == 'v1':
```

Comment:
A bit more descriptive naming would be better I think.
Or we could use an enum for the types.
```
class HashType:
    CONTEXT_FREE
    CONTEXT_SENSITIVE or PATH_SENSITIVE
```